frontend/views/containers/payments/PaymentsMixin.js

```diff
@@ -0,0 +1,97 @@
+import sbp from '@sbp/sbp'
+import { mapState, mapGetters } from 'vuex'
+import { PAYMENT_COMPLETED } from '@model/contracts/shared/payments/index.js'
+import { simplifyPayment, getPaymentHashes } from '@model/contracts/shared/functions.js'
+import { cloneDeep } from '@model/contracts/shared/giLodash.js'
+
+const PaymentsMixin: Object = {
+  computed: {
+    ...mapState(['currentGroupId']),
+    ...mapGetters([
+      'currentGroupState',
+      'ourUsername',
+      'ourPayments',
+      'paymentsForPeriod',
+      'periodBeforePeriod',
+      'groupPeriodPayments',
+      'currentPaymentPeriod',
+      'paymentHashesForPeriod',
+      'groupIncomeAdjustedDistribution'
+    ])
+  },
+  methods: {
+    async getHistoricalPaymentsInTypes () {
+      const paymentsInTypes = {
+        sent: cloneDeep(this.ourPayments?.sent),
+        received: cloneDeep(this.ourPayments?.received),
+        todo: this.groupIncomeAdjustedDistribution.filter(p => p.from === this.ourUsername)
+      }
+
+      const paymentsByPeriodKey = `paymentsByPeriod/${this.ourUsername}/${this.currentGroupId}`
+      const paymentsByPeriod = await sbp('gi.db/archive/load', paymentsByPeriodKey) || {}
+      const paymentsKey = `payments/${this.ourUsername}/${this.currentGroupId}`
+      const payments = await sbp('gi.db/archive/load', paymentsKey) || {}
+
+      for (const period of Object.keys(paymentsByPeriod).sort().reverse()) {
+        const { paymentsFrom } = paymentsByPeriod[period]
+        for (const fromUser of Object.keys(paymentsFrom)) {
+          for (const toUser of Object.keys(paymentsFrom[fromUser])) {
+            if (toUser === this.ourUsername) {
+              for (const rHash of paymentsFrom[fromUser][toUser]) {
+                const { data, meta } = payments[rHash]
+                paymentsInTypes.received.push({ hash: rHash, data, meta, amount: data.amount, username: data.toUser })
+              }
+            } else if (fromUser === this.ourUsername) {
+              for (const sHash of paymentsFrom[fromUser][toUser]) {
+                const { data, meta } = payments[sHash]
+                paymentsInTypes.sent.push({ hash: sHash, data, meta, amount: data.amount, username: data.toUser })
+              }
+            }
```

Comment:
This seems like it has some code duplication. We can DRY this more by doing something like this:

```js
if (toUser === this.ourUsername || fromUser === this.ourUsername) {
  const receivedOrSent = toUser === this.ourUsername ? 'received' : 'sent'
  for (const hash of paymentsFrom[fromUser][toUser]) {
      const { data, meta } = payments[hash]
      paymentsInTypes[receivedOrSent].push({ hash, data, meta, amount: data.amount, username: toUser })
    }
}
```

Reply:
Great solution.